### PR TITLE
add some filtering for the rest api endpoints to block all user endpo…

### PR DESCRIPTION
**Pull request by brandwaffle**
…ints by default. adds a filter that enables developers to further alter the array of allowed endpoints or place permissions checks to enable them conditionally

### CODE REVIEW

File: includes/rest-api.php [hunk 1]
```diff
@@ -1,6 +1,54 @@
...
+function rest_pre_dispatch($result, $rest_server, $request) {
+	$approved_routes = apply_filters( 'tenup_rest_approved_routes', $rest_server->get_routes() );
+
+	foreach( $approved_routes as $route => $data ) {
+		if( preg_match( '@^' . $route . '$@i', $request->get_route(), $matches ) ) {
+			return $result;
+		}
+	}
+
+	$result = 'your request is denied.';
```
Comment (pcrumm) [inline includes/rest-api.php]: Worth adding a more specific error message here to help anyone who isn't aware of this feature debug it? Perhaps something like `The 10up Experience plugin has disabled access to this endpoint.`

File: includes/rest-api.php [hunk 2]
```diff
@@ -1,6 +1,54 @@
...
+ * for example with a current_user_can() permission check
+ * @param $approved_routes The list of approved routes generated by WP_REST_Server::get_routes
+ * @return $approved_routes A filtered list of the routes that removes all user endpoints by default, for security.
+ */ 
+function create_default_routes( $approved_routes ) {
+	$endpoints_to_block = array(
+		'/wp/v2/users',
+		'/wp/v2/users/(?P<id>[\d]+)',
+		'/wp/v2/users/me',
+	);
```
Comment (pcrumm) [inline includes/rest-api.php]: The `posts` family is also a concern I see often around the WP community, though I'm not aware of any specific complaints we've received from clients regarding it. Still, I'm inclined to believe that we should either (a) block those endpoints or (b) treat this through a whitelist approach instead, and whitelist nothing by default. Thoughts?

File: includes/rest-api.php (hunk 1 above)
Comment (helen) [reply includes/rest-api.php]: I agree with @pcrumm about the contents of the message. I'm also wondering what the rest of the response looks like when using this filter - does it still 403 or is it a 200 or something else?

File: includes/rest-api.php [hunk 3]
```diff
@@ -1,6 +1,54 @@
...
+
+	foreach ( $approved_routes as $route => $data ) {
+		if ( in_array( $route, $endpoints_to_block ) ) {
```
Comment (tlovett1) [inline includes/rest-api.php]: We should pass `true` as the third param to `in_array` to force a strict comparison.

**Comment by helen** [comment]
Closing in favor of #15

**Pull request closed without merging**